content/Hash functions in daily life I - Martin Rubi.md

```diff
@@ -0,0 +1,172 @@
+at the time of leaving the laundry in the store and was printed in the ticket.
+
+The order number had around 10 digits digits.
+
+The last digit of each order number was a number between 0 and 9.
+
+That last digit could be used to optimize the lookup.
+
+To make the optimization the grid might be divided into different boxes.
+
+Some boxes would have a label assigned with a digit between 0 and 9.
+
+For example
+
+          -------------------------
+          |   |   |   |   |   |   |
+          -------------------------
+          | 7 | 8 | 9 | 0 |   |   |
+          -------------------------
+          | 1 | 2 | 3 | 4 | 5 | 6 |
+          -------------------------
+
+At the time of storing the laundry the cashier would look at the order number and
+take its last digit only. If the box corresponding to the last digit had empty
+space the bag would go into that box.
+
+If the box had not enough space for a new bag the bag would go to the first box
+with space among the unlabeled boxes.
+
+
+The lookup of a bag would apply the same algorithm. First she would look for the
+bag in the box corresponding to the last digit of the order. It the bag wasn't
+there she would look for it in the unlabeled boxes.
+
+That lookup is still O(N) in its worst case but in the average case it's O(1),
+meaning that, in average, the lookup would take only a constant number of comparisons.
+
+O(1) doesn't mean that it would take exactly one comparison. It means that the
+number of comparisons is a constant number and it does not depend on the total
+number of bags. In this case the constant number would be the number of bags that
+could fit in a single box.
+
+As an example, if a box had enough space for 4 bags the average lookup would take
+up to 4 comparisons regardless of the number of bags the rest of the shelf stored
+at that time.
+
+## Maps
+
+This algorithm is widely used in computer science and it's known as a hash table
+lookup.
+
+It's used to arrange items in such a way that a given item can be found in O(1)
+time.
+
+If you ever used a Map, a Dictionary or a Hash then you might have already used one.
+
+Map objects are commonly implemented as hash lookup tables.
+
+## Hash functions
+
+Taking the last digit from the order number is known as applying a hash function
+to the order number.
+
+A generalization of a hash function would take an infinite set, for instance the set
+of all integer numbers, and map each element to a finite set, for instance
+the numbers between 0 and 9.
+
+For this hash function to make an actual improvement on the search time the
+laundry order numbers would also need to meet some statistical conditions.
+
+The order number last digit would need to be evenly distributed among all orders.
+Otherwise the number of hash collisions would be too high and it would end up
+requiring a linear lookup on every search. Such a distribution is known as a
+Normal or Gaussian distribution.
```

Comment:
```suggestion
The order number last digit would need to be evenly distributed among all orders.
Otherwise, the number of hash collisions would be too high and it would end up
requiring a linear lookup on every search. Such a distribution is known as a
Normal or Gaussian distribution.
```